src/java.base/share/classes/sun/security/jca/JCAUtil.java

```diff
@@ -69,4 +71,25 @@ public static SecureRandom getSecureRandom() {
         return CachedSecureRandomHolder.instance;
     }
 
+    // called by sun.security.jca.Providers class when provider list is changed
+    static void clearDefSecureRandom() {
+        def = null;
+    }
+
+    /**
+     * Get the default SecureRandom instance. This method is the
+     * optimized version of "new SecureRandom()" which re-uses the default
+     * SecureRandom impl if the provider table is the same.
+     */
+    public static SecureRandom getDefSecureRandom() {
```

Comment:
When should people use `getSecureRandom` rather than `getDefSecureRandom()`? Are we leaving the old to avoid changing behavior out from under people who may not expect it?

Reply:
To minimize the impact, I leave the JCAUtil.getSecureRandom() impl as is. I suppose this is more for JDK internal code which is not required to use the most preferred SecureRandom impl. There are quite a few callers to this method and I feel it's better to leave them out of this change.

Reply:
> To minimize the impact, I leave the JCAUtil.getSecureRandom() impl as is. I suppose this is more for JDK internal code which is not required to use the most preferred SecureRandom impl. There are quite a few callers to this method and I feel it's better to leave them out of this change.

The internal use may be just to avoid to create new instances of SecureRandom.   Those use may need to use the latest updated secure random provider as well, for example the use in DSA key pair generation (DSAKeyPairGenerator.java).

With this update, I think the performance impact should be minimal and we may be able to have a uniform behavior no matter internal or external uses.

Reply:
Maybe. We can do that in a separate changeset...